lib/modules/datasource/rubygems/metadata-cache.ts

```diff
@@ -21,44 +32,61 @@ export class MetadataCache {
   ): Promise<ReleaseResult> {
     const cacheNs = `datasource-rubygems`;
     const cacheKey = `metadata-cache:${registryUrl}:${packageName}`;
-    const hash = toSha256(versions.join(''));
+    const versionsHash = hashVersions(versions);
 
-    const loadCache = (): AsyncResult<ReleaseResult, NonNullable<unknown>> =>
-      Result.wrapNullable(
+    const loadCache = (): AsyncResult<ReleaseResult, CacheError> =>
+      Result.wrapNullable<CacheRecord, CacheError, CacheError>(
         packageCache.get<CacheRecord>(cacheNs, cacheKey),
-        'cache-not-found' as const
+        'cache-not-found'
       ).transform((cache) => {
-        return hash === cache.hash
+        return versionsHash === cache.hash
           ? Result.ok(cache.data)
-          : Result.err('cache-outdated' as const);
+          : Result.err('cache-outdated');
       });
 
-    const saveCache = async (data: ReleaseResult): Promise<ReleaseResult> => {
+    const saveCache = async (
+      hash: string,
+      data: ReleaseResult
+    ): Promise<void> => {
       const registryHostname = parseUrl(registryUrl)?.hostname;
       if (registryHostname === 'rubygems.org') {
         const newCache: CacheRecord = { hash, data };
         const ttlMinutes = 100 * 24 * 60;
-        const ttlRandomDelta = Math.floor(Math.random() * 10 * 24 * 60);
+        const ttlDelta = 10 * 24 * 60;
+        const ttlRandomDelta = Math.floor(Math.random() * ttlDelta);
         await packageCache.set(
           cacheNs,
           cacheKey,
           newCache,
           ttlMinutes + ttlRandomDelta
         );
       }
-
-      return data;
     };
 
     return await loadCache()
-      .catch(() =>
-        getV1Releases(this.http, registryUrl, packageName).transform(saveCache)
-      )
-      .catch(() =>
-        Result.ok({
-          releases: versions.map((version) => ({ version })),
-        })
+      .catch((err) =>
+        getV1Releases(this.http, registryUrl, packageName).transform(
+          async (
+            newData: ReleaseResult
+          ): Promise<Result<ReleaseResult, CacheError>> => {
+            const v1ReleasesHash = hashReleases(newData);
```

Comment:
The key point of this PR: we don't use previously calculated hash from the `/versions` endpoint, we calculate it based on the `/api/v1/versions` response. This should make both cache layers consistent.